src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -545,7 +545,9 @@ fn build_base_args(cx: &Context,
     let prefer_dynamic = (unit.target.for_host() &&
                           !unit.target.is_custom_build()) ||
                          (crate_types.contains(&"dylib") &&
-                          unit.pkg.package_id() != &cx.current_package);
+                          cx.ws.current_opt().map_or(false, |p| {
+                              *p.package_id() != *unit.pkg.package_id()
+                          }));
```

Comment:
Wow I have absolutely no clue what this clause is doing (the old one).

Let's just replace it though with, like above, a test whether the package is in the current workspace.